### ISSUE TITLE
Manage the build version number better

### ISSUE THREAD

**Issue by craigfowler** [csf-dev/ZPT-Sharp#114]
The version number is currently managed from a number of separate places:
- In `Common/CommonAssemblyInfo.cs`
- The various `*.csproj` files
- In `CSF.Zpt.sln`
- In `Documentation/Doxygen.cfg`
- In the various `*.nuspec` files in the solution
- The version number passed to the documentation website build process

Currently I'm managing these locations by hand and making mistakes/omissions doing things one-by-one.

**Comment by craigfowler** [comment]
Perhaps it would be better to do this as part of the solution-wide pre-build procedure (as an msbuild task).

**Comment by craigfowler** [comment]
This could be handled after #137 - making it a pre-build step.

**Comment by craigfowler** [comment]
This might help with bumping the version number in the assembly info files:

http://www.msbuildextensionpack.com/help/4.0.5.0/html/d6c3b5e8-00d4-c826-1a73-3cfe637f3827.htm

**Comment by craigfowler** [comment]
I think that - whilst it would be nice to use the version from the solution file - it would be better to use one externally.

Then I can also deal with the incrementing of the build number (using semantic versioning).

The external file format would be:

* The current semantic version of the software
* The latest build number of that version.

This gives us: `X.Y.Z.b` where X, Y and Z indicate the semantic version and b is the build number which does not form part of the semantic version.

Then - the new build task can patch the semantic version number into the relevant files, but also the full version number into the assembly info file(s). This includes bumping the build number.

When changing version numbers we only change them in that external file, and we can manually reset the build number to zero.

**Comment by craigfowler** [comment]
Patching the version number into the doxygen config and the nuspec files should be handled by Deploy build steps. Not part of the pre build.

**Comment by craigfowler** [comment]
Difficult enough that I'm going to kick it down the road into sprint 7.

**Comment by craigfowler** [comment]
For assembly info versions, there is an msbuild extensions task which facilitates assembly info patching.

For doxygen, I would read the config from standard input. Then I can echo out an extra line for the version.
http://stackoverflow.com/a/11038467/6221779

For nuget it's possible to directly pass a version parameter to nuget pack.

**Comment by craigfowler** [comment]
In that case, there is no point worrying about the SLN or csproj files. The version info in them is worthless. Especially since visual studio support for those is patchy.

Maintain the version number in a text file; included the semantic version and also the build number in there. Maintain it by hand, everything else runs off of that.

**Comment by craigfowler** [comment]
# Plan for version number handling

## Files and source control
### Master version number
There will be a single new text file in a shared location named `SolutionVersion`. This will be checked into version control.

### Assembly info in projects
Assembly info files in each project will be replaced with a stub file, which will be located at both `AssemblyInfo.cs` and `AssemblyInfo.stub.cs`. Both will be checked into source control (but will never change). The stub version will be configured with the "do not compile" build property. The stub files basically have no content, just a comment to indicate that they are replaced at build-time.

### Skeleton assembly info
In a shared location, there will also be a skeleton `AssemblyInfo.skeleton.cs`. This has empty values for everything, but does contain all of the relevant attributes.

## Build tasks
Create two new build tasks:

* Read version number
* Increment build number

The **read version number** task reads the current semantic version number from a file and outputs it into a build property (variable).

The **increment build number** task reads the current seen fabric version number from a file and increments the fourth component (the build number).

This way, in pre-build, the build number can be incremented, then every subsequent project's build step can just read it.

## Build sequence
### Incrementing the version
For a debug build, the version number is never incremented.

In a release or deploy build, one of the first things to happen is that the version number is incremented.

### Copying a skeleton AssemblyInfo
A pre-build step for all projects will be to copy the skeleton assembly info file into the project's directory, over the top of the assembly info there.

### Populating AssemblyInfo
Also in pre-build, the assembly info task runs, which populates information in the newly-copied assembly info file, using configuration data in the project file and in the version number file.

### Restoring the stub assembly info
As a post-build task, the stub assembly info full he is copied back over the patched one, restoring it to its original state.

### Creating Nuget packages
Pass the version number in by hand on the command line, ignore the version number in the package itself.

*This needs the semantic version and not the assembly version.*

### Doxygen config
Use the STDIN trick to set the version for that at run-time as well. Omit the version parameter from the config.

*This needs the semantic version and not the assembly version.*
